File: roms/flicky.py
#
#	Flicky
#

from base64 import b64encode
from io import BytesIO
from math import ceil
from sys import argv
from zipfile import ZipFile
from PIL import Image

with ZipFile(argv[1]) as z:
    prg1 = z.read('flickyo/epr-5857.bin') + z.read('flickyo/epr-5858a.bin') + z.read('flickyo/epr-5859.bin') + z.read('flickyo/epr-5860.bin')
    prg2 = z.read('epr-5869.120')
    bg = z.read('epr-5868.62') + z.read('epr-5867.61') + z.read('epr-5866.64') + z.read('epr-5865.63') + z.read('epr-5864.66') + z.read('epr-5863.65')
    obj = z.read('epr-5855.117') + z.read('epr-5856.110')
    pri = z.read('pr-5317.76')

rom = prg1 + prg2 + bg + obj + pri

def pngstring(a):
    w = 1024
    img = Image.new('L', (w, ceil(len(a) / w)))
    img.putdata(a)
    buf = BytesIO()
    img.save(buf, 'PNG')
    return b64encode(buf.getvalue())

with open(argv[2], 'wb') as f:
    f.write(b'export const ROM = \'data:image/png;base64,' + pngstring(rom) + b'\';\n')

File: roms/flicky.js
/*
 *	Flicky
 */

const fs = require('fs');
const {createCanvas, createImageData} = require('canvas');
const {Zlib} = require('zlibjs/bin/unzip.min');
Zlib.Unzip.prototype.read = Zlib.Unzip.prototype.decompress;

const cat = (...args) => {
	const array = new Uint8Array(args.reduce((a, b) => a + b.length, 0));
	return args.reduce((a, b) => (array.set(b, a), a + b.length), 0), array;
};

const z = new Zlib.Unzip(fs.readFileSync(process.argv[2]));
const prg1 = cat(z.read('flickyo/epr-5857.bin'), z.read('flickyo/epr-5858a.bin'), z.read('flickyo/epr-5859.bin'), z.read('flickyo/epr-5860.bin'));
const prg2 = z.read('epr-5869.120');
const bg = cat(z.read('epr-5868.62'), z.read('epr-5867.61'), z.read('epr-5866.64'), z.read('epr-5865.63'), z.read('epr-5864.66'), z.read('epr-5863.65'));
const obj = cat(z.read('epr-5855.117'), z.read('epr-5856.110'));
const pri = z.read('pr-5317.76');

const rom = cat(prg1, prg2, bg, obj, pri);

const pngString = array => {
	const w = 1024, h = Math.ceil(array.byteLength / w), canvas = createCanvas(w, h);
	const palette = new Uint8ClampedArray(new Uint32Array(256).map((e, i) => i | 0xff000000).buffer);
	canvas.getContext('2d', {pixelFormat: 'A8'}).putImageData(createImageData(new Uint8ClampedArray(array.buffer), w, h), 0, 0);
	return Buffer.from(canvas.toBuffer('image/png', {compressionLevel: 9, filters: canvas.PNG_FILTER_NONE, palette})).toString('base64');
};

fs.writeFileSync(process.argv[3], `export const ROM = 'data:image/png;base64,${pngString(rom)}';\n`);
